### ISSUE TITLE
Tracking Issue: PHP 8 Support

### ISSUE THREAD

**Issue by czosel** [prettier/plugin-php#1691]
Released with `v0.17.0`:

- [x] nullsafepropertylookup
- [x] named arguments
- [x] match expression
- [x] union types
- [x] promoted properties

Released with `v0.17.4`:

- [x] attributes

---
8.1 Support:
===

- [x]  `enum`:
```
enum Suit {
    case Clubs;
    case Diamonds;
    case Hearts;
    case Spades;
}
```

- [x] readonly class parameters
```php
class ImutableData {
    public function __construct(
        public readonly string $key,
    ) {}
}
```
- [x] first class callable syntax - eg `array_filter($in, trim(...));`
- [x] Intersection Types in function parameters:
```php
function generateSlug(HasTitle&HasId $post) {
    return strtolower($post->getTitle()) . $post->getId();
}
```
- [x] never return type
- [x] Octal Integer Notation `0o` or `0O`
- [x] [new in initializers](https://wiki.php.net/rfc/new_in_initializers)

**Comment by Selion05** [comment]
So attributes is the last one?  Yesterday I started working on it, but it's tricky!

**Comment by czosel** [comment]
Attributes is the last one without a first implementation in the parser. Match, union and promoted props still need implementation in our prettier plugin.
I'd love to help, but I'm short on time :disappointed:

**Comment by Selion05** [comment]
Ok cool. I'll see what I can do tomorrow

**Comment by cseufert** [comment]
We are now only waiting on Attributes support to land full PHP8 syntax support.

**Comment by czosel** [comment]
A large part of the new features just landed in `v0.17.0` :tada: The only thing that's still open is support for attributes. Please try the new release and report back if you run into issues :slightly_smiling_face:

**Comment by markkirchner** [comment]
Specifying `"phpVersion": "8.0"` in `.prettierrc.json` does not work (fails with `Invalid phpVersion value`). So when testing the new release I should stick to "7.4"?

**Comment by cseufert** [comment]
I suspect you have not actually upgraded to 0.17, if you leave the level at 7.4 then prettier will fail to parse any files using PHP 8.0 languages features (eg. match statement, union types, property promotion, etc).

**Comment by markkirchner** [comment]
> I suspect you have not actually upgraded to 0.17

Thanks for the hint. In fact, I already had upgraded to 0.17, but the changes in `node_modules` had not been synced to the machine running prettier... ☹️
Works now as expected, sorry for the noise.

**Comment by alexander-akait** [comment]
updated about enum

**Comment by cseufert** [comment]
> updated about enum

Added the rest of the syntax changing PHP8.1 features

**Comment by chrisrhymes** [comment]
Is there an option to force prettier to ignore attributes until support is added? 

I tried the following on a class property:

```php
// prettier-ignore
#[CastWith(CarbonCaster::class)]
public Carbon $startDate;
```

But when running the `--check` command it returns the following with the attribute removed and the `public` removed too:
```php
// prettier-ignore
Carbon $startDate
```

Thanks.

**Comment by cseufert** [comment]
@chrisrhymes the main issue is that the parser parses the attributes, and prettier currently has no code to turn the attributes that have been parsed, back into output.

**Comment by cseufert** [comment]
@chrisrhymes Are you able to update to 0.17.4, it should now be capable of formatting the attributes. Please open a new issue for any weird behaviour.

**Comment by chrisrhymes** [comment]
Thank you @cseufert . Tested 0.17.4 in my project and all works as expected!

**Comment by Selion05** [comment]
Do you pass`8.0` as a string or a number? It is supposed to be a string. You can also try as a number `800`

**Comment by elev8studio** [comment]
> Do you pass`8.0` as a string or a number? It is supposed to be a string. You can also try as a number `800`

I'm passing as a string 
```
{
  "phpVersion": "8.0",
  ...
}
```

I also tried passing as a number. I get the same result.

**Comment by elev8studio** [comment]
The solution was to upgrade the main `prettier` package to the latest version. It works now!

**Comment by elev8studio** [comment]
The exact error I get is this: `Error: Invalid phpVersion value. Expected "5.0", "5.1", "5.2", "5.3", "5.4", "5.5", "5.6", "7.0", "7.1", "7.2", "7.3" or "7.4", but received "8.0".` It seems that `"8.0"` is not even an acceptable option! Please, what's the solution?

**Comment by alexander-akait** [comment]
run `npm ls @prettier/plugin-php`

**Comment by elev8studio** [comment]
This is the output 
```
└─ @prettier/plugin-php@0.17.4
✨  Done in 0.52s.
```
It's definitely the right version.

**Comment by cseufert** [comment]
Perhaps you also have it installed globally? try removing the global installation.
```sh
npm uninstall -g @prettier/plugin-php prettier
```

**Comment by tm1000** [comment]
Enum support was added by @czosel yesterday https://github.com/glayzzle/php-parser/pull/862

**Comment by genintho** [comment]
PR adding enum support this plugin https://github.com/prettier/plugin-php/pull/1916

**Comment by genintho** [comment]
Support for 50% of the readonly properties syntax  https://github.com/prettier/plugin-php/pull/1917 
Adding support to the syntax involving the class constructor will require a much bigger change in the code of the php-parser for what I can tell.

**Comment by czosel** [comment]
Readonly support landed in `0.18.2` :tada:

**Comment by genintho** [comment]
PR adding Intersection types support https://github.com/prettier/plugin-php/pull/1933.
Requires parser modification https://github.com/glayzzle/php-parser/pull/883

**Comment by mzkiskw** [comment]
I specify `phpVersion: "8.1"` in `prettierrc.js`, I get this is:
```
[error] Invalid phpVersion value. Expected "5.0", "5.1", "5.2", "5.3", "5.4", "5.5", "5.6", "7.0", "7.1", "7.2", "7.3", "7.4" or "8.0", but received "8.1".
```

```bash
$ yarn list | grep @prettier/plugin-php
├─ @prettier/plugin-php@0.18.1
```

Please can you help?

**Comment by mzkiskw** [comment]
Sorry, after I run `yarn upgrade @prettier/plugin-php`, that error is gone.

**Comment by thomask** [comment]
Should "[new in initializers](https://wiki.php.net/rfc/new_in_initializers)" be added to the list of Todos? This is currently throwing an error for me:

````
[error] src/Entity/Foo.php: SyntaxError: Parse Error : syntax error, unexpected 'ORM\Ass...' (T_STRING), expecting ']' on line 18
[error]   16 | ])]
[error]   17 | #[ORM\AssociationOverrides(overrides: [
[error] > 18 |     new ORM\AssociationOverride(name: 'foo', inversedBy: 'bars')
[error]      |        ^
[error]   19 | ])]
````

**Comment by hackel** [comment]
Prettier is removing trailing commas from match expressions when `trailingCommaPHP=true`.

Wasn't sure if you wanted a separate issue for this or just to track it here.

**Comment by czosel** [comment]
@hackel thanks for the bug report - I guess a new issue would be better, so I went ahead and created one.

**Comment by tm1000** [comment]
@thomask FYI @czosel added "[new in initializers](https://wiki.php.net/rfc/new_in_initializers)" to the TODO! (I have the same issue as you).

Thanks for your work thus far @czosel

**Comment by AndreasA** [comment]
Any news regarding first class callable syntax ?

**Comment by AndreasA** [comment]
Just wondering but I think the new in initializers might actually already work too? At least it seems to work on the playground: https://loilo.github.io/prettier-php-playground/#N4IgDgTgpgLjCWUIHkwIPYDsDOIBcok8mMA6vACYwAW+AHAAwA0IMAhgEblW14AsLAK7YoAFU648AMzYAbES2zEA5rKgBFQehhR8M+VBZhqYAGpIlWfCDoA6AIwgWMCG3iyVAYXQBbH2wAFAAkA-BdBQxAOVwBjKABlGABPNWswbAgAWgAmJxBoAEdBeGgA12V-PTkFEGIRCBgytgq2KoMAXxYoCngYdBQ0eCxJUDZsGDaRTtrMMEEJvBAAHgB+YzAAAgooGNk2aAAKcYh4GJgAfWSwKGwAXnsASgBuAB1MN8w2HxuwNjiNgBi6HQr3emGEUA2ACF9i8XgA5QSyWQAGXQymUSFBb12Y2wG3iSAAbqcoG9gG8NlSNkQiWwdBs0RikBsACSydGYiCg6mU6lzDgeGIbKSCTBnIaYDbnc4xYbhM4HPnU6lMrlsjnMiAbW4bTBQADuG0RyLVSAODyYypVNqpcJhEGycLhZu17M5LNu1qp+qN9v2ToRSNRHogFqtUttUddGtDOr1ho23uNwdd4etDw2FMjttZNHg2EyAD5Nerde6tTzqe03u0QO0gA

**Comment by czosel** [comment]
@AndreasA You're right, looks like this works already! :+1: I'm also not sure in which context this was fixed.

**Comment by tm1000** [comment]
Agree new in initializers works! yay!

I guess the *only* thing left is `Octal Integer Notation 0o or 0O`

**Comment by AndreasA** [comment]
That and maybe some minor formatting improvements like https://github.com/prettier/plugin-php/issues/1870 but that isn't that important as general support.

@czosel So I guess as it is now confirmed by someone else as well that the corresponding point above https://github.com/prettier/plugin-php/issues/1691#issue-849676387 can be marked as done?

**Comment by czosel** [comment]
Octal integer notation landed in `v0.18.9` :tada:

**Comment by tm1000** [comment]
@czosel seems like you got everything done on this! you can probably close it now yeah?
